cpp/src/arrow/engine/substrait/extension_set.cc

```diff
@@ -645,22 +663,19 @@ struct ExtensionIdRegistryImpl : ExtensionIdRegistry {
 };
 
 template <typename Enum>
-using EnumParser = std::function<Result<Enum>(std::optional<std::string_view>)>;
+using EnumParser = std::function<Result<Enum>(std::string_view)>;
 
 template <typename Enum>
 EnumParser<Enum> GetEnumParser(const std::vector<std::string>& options) {
   std::unordered_map<std::string, Enum> parse_map;
   for (std::size_t i = 0; i < options.size(); i++) {
     parse_map[options[i]] = static_cast<Enum>(i + 1);
```

Comment:
It's probably not necessary to reify "UNSPECIFIED" in all the enumerations. We will always either coerce to a default or error because a real value was required. That could be expressed in EnumParser as:
```c++
template <typename Enum>
using EnumParser = std::function<Result<Enum>(std::string_view repr, std::optional<Enum> default)>;
```
... and it would save you some one-based indexing headache.

Probably out of scope for this PR

Reply:
Agreed to leave out of scope for this PR.  I also agree it would probably be a good idea.  The "unspecified" thing is a relic from the old behavior where "unspecified" was something you had to explicitly specify.